content/blog/2023-12-09-hcl-amc/index.md

```diff
@@ -0,0 +1,259 @@
+
+The [CIRCT](https://circt.llvm.org/) project expands the MLIR-based development methodology for hardware design. It represents key components of hardware as MLIR dialects such as finite state machines (FSM), pipelines, and interface handshaking. HeteroCL has been migrated to the MLIR ecosystem as a dialect, with a new Python frontend called HCL. HCL decouples the interactions between the algorithm, hardware optimizations, and backend targets to enable productive design and testing. Lastly, Accelerator Memory Compiler (AMC) is an MLIR dialect for representing memory architecture. Its expressiveness is able to capture common memory organization strategies such as partitioning, banking, and arbitration. AMC can be further lowered to Calyx, which is also part of the CIRCT ecosystem. Calyx IR gives us a pathway to finally to synthesizable Verilog. The contribution of this project is that we integrated HCL with AMC to enable a Python frontend for AMC. This allows us to use HCL to describe the algorithm and AMC to describe the memory architecture. The resulting design can be compiled to Verilog and simulated with a single function call. In the end, we hope that this integration will enable a more productive design flow for hardware accelerators as well as help us find more bugs in AMC.
+
+## Design Example
+
+To use HCL with AMC, the designer only needs to write their kernel in Python. Then, then user can simply specify which Python function to build with the AMC backend. Moreover, AMC acts as a drop-in replacement to the other backends in the HCL ecosystem, making functional testing and debugging seamless. In the far majority of cases, the [NumPy](https://numpy.org/) or the LLVM backend is suitable for use as a golden reference model. In this section, we walk through an example where we functionally verify a kernel built with AMC. Then, we will record some resource estimates and execution times.
+
+Our illustrative example will be matrix multiplication. What would ordinarily be a cumbersome task when using the vendor tools, like [Vitis HLS](https://www.xilinx.com/products/design-tools/vitis/vitis-hls.html), becomes a simple, 5 minute exercise. First, we specify some inputs initialized to random values. Then, `build()` the accelerator. Finally, we call both the software and hardware simulations and check their outputs. Compared to a C/C++ based tool flow, the amount of boilerplate code and scripting is reduced to near zero. In the end, we can represent this application with only 18 lines of code:
+
+```python
+def test_amc():
+    N = 16
+    # Initialize 2 input matricies to random integers
+    A = np.random.randint(0, 20, size=(N, N), dtype="int32")
+    B = np.random.randint(0, 20, size=(N, N), dtype="int32")
+
+    # Our kernel is just matrix multiplication
+    # HCL provides a primitive for this
+    def kernel(A: int32[N, N], B: int32[N, N]) -> int32[N, N]:
+        return matmul(A, B)
```

Comment:
Do you think you could add something quick here about how the Python frontend works (and exactly which component is responsible for it)? Namely, I think the answers are:

* It works by statically inspecting the Python AST and directly translating a subset of that to the `affine` dialect. (That is, it does _not_ work with the dynamic operator-overloading approach that some other Python frontends do.)
* The component that does this Python AST translation is either part of HCL itself… or perhaps it is from something external like the [npcomp](https://github.com/google/mlir-npcomp/tree/archived) project?

Understanding exactly who/what does this translation and how it works will help the reader see what the power and limitations are here.